loki/transformations/single_column/vertical.py

```diff
@@ -0,0 +1,218 @@
+            return
+        role = kwargs['role']
+        if role == 'kernel':
+            self.process_kernel(routine)
+
+    def process_kernel(self, routine):
+        """
+        Current logic (simplified):
+
+        1. loop interchange to expose vertical loops
+        2. fuse vertical loops (possibly into multiple groups)
+        3. find local arrays to be demoted and apply heuristics to check whether this is safe
+        4. demote those arrays which are safe to be demoted
+        """
+        # find local arrays with a vertical dimension
+        relevant_local_arrays = self.find_relevant_local_arrays(routine)
+        # find "multilevel" thus "jk +/- 1" arrays
+        multilevel_relevant_local_arrays = self.identify_multilevel_arrays(relevant_local_arrays)
+        # loop interchange to expose vertical loops as outermost loops
+        loop_interchange(routine)
+        # handle initialization of arrays "jk +/- 1" arrays
+        multilevel_relevant_local_arrays_names = set(arr.name.lower() for arr in multilevel_relevant_local_arrays)
+        self.correct_init_of_multilevel_arrays(routine, multilevel_relevant_local_arrays_names)
+        # fuse vertical loops
+        loop_fusion(routine)
+        # demote in vertical dimension if possible
+        relevant_local_arrays_names = set(arr.name.lower() for arr in relevant_local_arrays)
+        demote_candidates = relevant_local_arrays_names - multilevel_relevant_local_arrays_names
+        # check which variables are safe to demote in the vertical
+        safe_to_demote = self.check_safe_to_demote(routine, demote_candidates)
+        # demote locals in vertical dimension
+        dimensions_to_demote = self.vertical.size_expressions + (f"{self.vertical.size}+1",)
+        demote_variables(routine, safe_to_demote, dimensions_to_demote)
+
+    def check_safe_to_demote(self, routine, demote_candidates):
+        """
+        Check whether variables that are candidates to be demoted in the vertical dimension are really
+        safe to be demoted.
+
+        Current heuristic: If the candidate is used in more than one vertical loop, assume it is NOT safe
+        to demote!
+        """
+        loop_var_map = {}
+        with pragmas_attached(routine, ir.Loop):
+            for loop in FindNodes(ir.Loop).visit(routine.body):
+                if loop.variable.name.lower() == self.vertical.index.lower():
+                    ignore = False
+                    if is_loki_pragma(loop.pragma, starts_with='fused-loop'):
+                        parameters = get_pragma_parameters(loop.pragma, starts_with='fused-loop')
+                        group = parameters.get('group', 'default')
+                        if group == 'ignore':
+                            ignore = True
+                    if not ignore:
+                        loop_var_map[loop] = as_tuple(var.name.lower() for var in FindVariables().visit(loop.body)
```

Comment:
Would building the inverse of this map not make the logic much simple? As in `var_loop_map[var.name] = loop` would allow you to then simply check for `len(var_loop_map[var.name]) > 1` in the next part? Or am I misreading this? 

Also, a `CaseInsensitiveDict` might be useful here for `var_loop_map`.